cloud/terraform/machineactuator.go

```diff
@@ -351,10 +353,110 @@ func (tf *TerraformClient) PostDelete(cluster *clusterv1.Cluster, machines []*cl
+		goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
+		goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+
+		// If the saved versions and new versions differ, do in-place upgrade.
+		if goalMachineControlPlaneVersion != goalMachine.Annotations[ControlPlaneVersionAnnotationKey] ||
+			goalMachineKubeletVersion != goalMachine.Annotations[KubeletVersionAnnotationKey] {
+			glog.Infof("Doing in-place upgrade for master from %s to %s", goalMachine.Annotations[ControlPlaneVersionAnnotationKey], goalMachineControlPlaneVersion)
+			err := tf.updateMasterInPlace(goalMachine)
+			if err != nil {
+				glog.Errorf("Master inplace update failed: %+v", err)
+			}
+		} else {
+			glog.Info("UPDATE TYPE NOT SUPPORTED")
+			return nil
+		}
+	} else {
+		glog.Info("NODE UPDATES NOT SUPPORTED")
+		return nil
+	}
+
+	return nil
+}
+
+// Assumes that update is needed.
+// For now support only K8s control plane upgrades.
+func (tf *TerraformClient) updateMasterInPlace(goalMachine *clusterv1.Machine) error {
+	goalMachineControlPlaneVersion := goalMachine.Spec.Versions.ControlPlane
+	goalMachineKubeletVersion := goalMachine.Spec.Versions.Kubelet
+
+	currentControlPlaneVersion := goalMachine.Annotations[ControlPlaneVersionAnnotationKey]
+	currentKubeletVersion := goalMachine.Annotations[KubeletVersionAnnotationKey]
+
+	// Control plane upgrade
+	if goalMachineControlPlaneVersion != currentControlPlaneVersion {
+		// Pull the latest kudeadm
+		cmd := fmt.Sprintf("curl -sSL https://dl.k8s.io/release/%s/bin/linux/amd64/kubeadm | sudo tee /usr/bin/kubeadm > /dev/null; " +
+			"sudo chmod a+rx /usr/bin/kubeadm", goalMachineControlPlaneVersion)
+		_, err := tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while downloading new kubeadm: %+v", err)
+			return err
+		}
+
+		// Next upgrade control plane
+		cmd = fmt.Sprintf("sudo kubeadm upgrade apply %s -y", "v"+goalMachine.Spec.Versions.ControlPlane)
+		_, err = tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand failed while upgrading control plane: %+v", err)
+			return err
+		}
+	}
+
+	// Upgrade kubelet
+	if goalMachineKubeletVersion != currentKubeletVersion {
+		// Upgrade kubelet to desired version.
+		cmd := fmt.Sprintf("sudo apt-get install kubelet=%s", goalMachine.Spec.Versions.Kubelet+"-00")
+		_, err := tf.remoteSshCommand(goalMachine, cmd, "~/.ssh/id_rsa", "ubuntu")
+		if err != nil {
+			glog.Infof("remoteSshCommand while installing new kubelet version: %v", err)
+			return err
+		}
+	}
+
 	return nil
 }
 
+func (tf *TerraformClient) remoteSshCommand(m *clusterv1.Machine, cmd, privateKeyPath, sshUser string) (string, error) {
+	glog.Infof("Remote SSH execution '%s' on %s", cmd, m.ObjectMeta.Name)
+
+	publicIP, err := tf.GetIP(m)
+	if err != nil {
+		return "", err
+	}
+
+	command := fmt.Sprintf("echo STARTFILE; %s", cmd)
```

Comment:
I didn't see that in the review thread. I created https://github.com/kubernetes-sigs/cluster-api/issues/164 for tracking this and will handle it in a separate PR since it affects all existing controllers.